### DISCUSSION TITLE
Event bus

### DISCUSSION THREAD

**Discussion by vzahradnik** [davetcc/tcMenu#369] [Ideas]
While I'm thinking of a more decoupled architecture for my firmware using TaskManager from TcMenu, I figured that it would be useful to implement an event bus.

For example, once the device is connected to the network, it will broadcast a message. Other components of the app will listen to this event and can start an activity like initializing an MQTT client.

I believe it shouldn't be difficult to implement and I also think the TcMenu ecosystem would benefit from a feature like this.

**Comment by davetcc** [comment]
Agreed, I think this would be a good addition. It was in the original plan to support MQQT and other decoupled solutions.

**Reply by vzahradnik** [reply]
I didn't do any research. If there is something we could integrate, great! Otherwise I was thinking about a simple publisher-subscriber model with a queue. You already have an implementation of a cyclic buffer, right?

**Reply by davetcc** [reply]
Yep, there should be everything we need to implement this right in the core, using TaskManager's event support along with a cyclic buffer.

Just remember that MenuItems are not thread safe, so if they are to cross cores or be used in a raw interrupt, they should be copied, as they are only small, the overhead should be minor. The info blocks only need to be copied in the event that they are in RAM instead of FLASH, if they're in FLASH, they are effectively immutable. We could add an "=default" copy constructor and equals operator to each menu item that handled the item copy, and we could provide a copy wrapper that could be stored in the circular buffer, that did the whole lot transparently.

Given this is only likely to be used on larger boards, the overhead would be negligible. A few KB extra basically.

**Reply by vzahradnik** [reply]
Threads are an issue only when using mbed, I suppose. Anyway, let's do it right.

**Reply by davetcc** [reply]
If everything is in task manager, there is no issue with threads on any board, but if the circular buffer were to be used on another core, let's say to offload work (to another task on ESP32, or thread in mbed), then we'd need to think about that. but that's probably a longer-term consideration.

**Reply by vzahradnik** [reply]
Whoever decides to use an event bus has a more complex firmware and hence a more powerful board. So memory shouldn't be an issue. I use ESP32.

**Comment by davetcc** [comment]
There are issues for most things raised here. I'll close this discussion
